proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,671 @@
+  receive :: Socket Connected -> IOL (Socket Connected, Unrestricted ByteString)
+  close :: ∀s. Socket s -> IOL ()
+
+The `paper <https://arxiv.org/abs/1710.09756>`_ mentions other use
+cases as well, such as efficient and safe data serialization as well
+as using types to statically enforce interaction protocols between
+communicating processes (*e.g.* via RPC calls). GHC users will no
+doubt invent many more use cases over time.
+
+Proposed Change Specification
+-----------------------------
+
+We introduce a new language extension. Types with a linearity
+specification are syntactically legal anywhere in a module if and only
+if ``-XLinearTypes`` is turned on.
+
+Definition
+~~~~~~~~~~
+
+We say that a function ``f`` is *linear* when ``f u`` is consumed
+exactly once implies that ``u`` is *consumed exactly once* (defined
+as follows).
+
+- Consuming a value of a data type exactly once means evaluating it to
+  head normal form, then consuming its fields exactly once
+- Consuming a function exactly once means applying it and consuming
+  its result exactly once
+
+The type of linear function from type ``A`` to type ``B`` is written
+``A ->. B`` (see syntax below).
+
+Linearity is a strengthening of the contract of the regular function
+type ``A -> B``, which will be called the type of *unrestricted*
+functions.
+
+Remark: linear function ``f`` can diverge or be called on diverging
+data. It may feel weird because ``f`` will not necessarily consume its
+argument. But it's alright: the metatheory still works. One can think
+of a diverging computation as not consuming its result exactly once,
+therefore, it is vacuously true that if ``f u`` is consumed exactly
+once, then ``u`` is consumed exactly once.
+
+Polymorphism
+~~~~~~~~~~~~
+
+In order for linear functions and unrestricted functions not to live
+in completely distinct worlds, to avoid code duplication, we
+introduce a notion of polymorphism, dubbed *multiplicity polymorphism*,
+over whether a function is linear.
+
+A linear function is said to have multiplicity ``1`` while an
+unrestricted function is said to have multiplicity ``ω``. Multiplicity
+polymorphic functions may have variable multiplicity, *e.g.*
+
+::
+
+  map :: (a:p -> b) -> [a]:p -> [b]
+
+without polymorphism we would need two implementations of `map`. With
+the exact same code: one for ``p=1`` and one for ``p=ω``. Function
+composition is even worse: it takes two multiplicity parameters, hence,
+would require four identical implementations:
+
+::
+
+  (.) :: (b:p -> c) -> (a:q -> b) -> a:(p ':* q) -> c
+
+Syntax
+~~~~~~
+
+The new primary constructs are: multiplicities and the multiplicity
+indexed arrow.
+
+- Multiplicities are a datatype:
+
+  ::
+
+    data Multiplicity
+      = One
+      | Omega
+      | Multiplicity :* Multiplicity
+      | Multiplicity :+ Multiplicity
+
+  In the following, for conciseness, we write ``1`` for ``One`` and
+  ``U`` (ASCII) or ``ω`` (Unicode) for ``Omega``. Note: unification
```

Comment:
It would make sense to stick to what one can write in the language. If the programmer must write `a :'One-> b` it doesn't make sense to have the proposal hide the fact with a notation for conciseness (a :'1-> b).